### DISCUSSION TITLE
Difference between DispatcherTimer and DispatcherQueueTimer

### DISCUSSION THREAD

**Discussion by fabianoriccardi** [microsoft/WindowsAppSDK#4770] [Q&A] [answered]
Which is the difference between DispatcherTimer and DispatcherQueueTimer? 

Reading the docs, it seems they are different api to obtain the same result (I.e. running code on a specified dispatcherqueue/thread. If so, why this duplication?

My concrete use case is the simplest one: running few lines on the UI thread.

**Comment by ghost1372** [comment]


**Reply by fabianoriccardi** [reply]
You (or Copilot, as you prefer) didn't reply to my whole question. 

To run few (light) lines of codes on UI thread, what do the developers suggest? One of the 2 API will be deprecated? Is there a preferred path or both approaches are the same for this specific purpose? 

I have to add a detail: I'm focusing on WinUI3, no WPF.

Reading the docs, I was supposing the best is DispatcherQueueTimer, which is more flexible and can do all the tasks of DispatherTimer. Again, I'm not sure of this guess.

Finally, if there is a preference among the 2 API, I think this should be reported in the documentation.

**Reply by DarranRowe** [reply]
@ghost1372 
[Microsoft.UI.Xaml.DispatcherTimer](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.dispatchertimer?view=windows-app-sdk-1.6) exists.

**Comment by DarranRowe** [comment, marked as answer]
Essentially, it is historical and on the UWP side where the Xaml interface was copied from.

Basically, before the DispatcherQueue was added to Windows, the only dispatcher available for UWP applications was the CoreDispatcher. Back then, DispatcherQueueTimer also didn't exist. I hope it is obvious why. The only UI thread timer available was the DispatcherTimer.

Now, when reimplementing Xaml for the desktop environment, changes had to be made to work with that transition. One important difference between the desktop and UWP environments is that the desktop environment doesn't have a CoreDispatcher running. Only a DispatcherQueue is available, which was also reimplemented in Microsoft.UI.Dispatching. So what does this mean for the DispatcherTimer? Well, in order to ease porting, it was reimplemented but to depend on the DispatcherQueue.

As to which one to use, they should be pretty similar, I'm sure DispatcherTimer is essentially an alias of DispatcherQueueTimer, where the "constructor" would basically call DispatcherQueue.GetForCurrentThread().CreateTimer(). However, the DispatcherQueueTimer is more flexible, so that would be a better option.

**Reply by fabianoriccardi** [reply]
This is not the first time you give me a complete and exhaustive reply, thanks :)

**Reply by hawkerm** [reply]
You should always use `DispatcherQueue` and `DispatcherQueueTimer` over `Dispatcher` and `DispatcherTimer`, they're not forwarded to my knowledge and usually null in the WinUI context. You're right it's a historical from UWP time and the reimplementation in WinUI, but they were added in the OS ahead of time to have a dispatcher not tied to the window.

The only place you should see the plain `Dispatcher` is in old UWP apps. For any new development, grab a reference to the `DispatcherQueue` or `DispatcherQueueTimer` in your UI Element's constructor in order to have a reference to the UI thread, use it elsewhere in order to dispatch back to the UI thread.
